ntpd/src/daemon/server.rs

```diff
@@ -196,321 +283,262 @@ impl<C: 'static + NtpClock + Send> ServerTask<C> {
-                packet,
-                max_response_size,
-                decoded_cookie,
-                peer_addr,
-                recv_timestamp,
-            } => {
-                self.stats.accepted_packets.inc();
-
-                let keyset = self.keyset.borrow().clone();
-                let mut buf = [0; MAX_PACKET_SIZE];
-                let mut cursor = Cursor::new(buf.as_mut_slice());
-                let serialize_result = match decoded_cookie {
-                    Some(decoded_cookie) => {
-                        self.stats.nts_received_packets.inc();
-                        self.stats.nts_accepted_packets.inc();
-                        let response = NtpPacket::nts_timestamp_response(
-                            &self.system,
-                            packet,
-                            recv_timestamp,
-                            &self.clock,
-                            &decoded_cookie,
-                            &keyset,
-                        );
-                        response.serialize(&mut cursor, decoded_cookie.s2c.as_ref())
-                    }
-                    None => {
-                        let response = NtpPacket::timestamp_response(
-                            &self.system,
-                            packet,
-                            recv_timestamp,
-                            &self.clock,
-                        );
-                        response.serialize(&mut cursor, &NoCipher)
+                // For a server, we only trigger NetworkGone restarts
+                // on ENETDOWN. ENETUNREACH, EHOSTDOWN and EHOSTUNREACH
+                // do not signal restart-worthy conditions for the a
+                // server (they essentially indicate problems with the
+                // remote network/host, which is not relevant for a server).
+                // Furthermore, they can conceivably be triggered by a
+                // malicious third party, and triggering restart on them
+                // would then result in a denial-of-service.
+                match receive_error.raw_os_error() {
+                    Some(libc::ENETDOWN) => SocketConnection::Reconnect,
+                    _ => {
+                        self.stats.ignored_packets.inc();
+                        SocketConnection::KeepAlive
                     }
-                };
-
-                if let Err(serialize_err) = serialize_result {
-                    warn!(error=?serialize_err, "Could not serialize response");
-                    return true;
                 }
+            }
+            Ok((length, peer_addr, opt_timestamp)) => {
+                let Some(request_buf) = buf.get(..length) else {
+                    warn!("length from socket is out of bounds. This is a bug!");
+                    return SocketConnection::Reconnect;
+                };
+                let mut response_buf = [0; MAX_PACKET_SIZE];
+                let response_buf = response_buf.as_mut_slice();
```

Comment:
We should perhaps limit the response buffer size here already.

Reply:
Great idea, changed it and removed the old check at the end. So now if this ever happens it should just show up as a serialization error.